### PR TITLE
Changed fit and transform functions in multiscale_phate.py

### PR DESCRIPTION
**Pull request by mkuchroo**

### CODE REVIEW

File: multiscale_phate/compress.py [hunk 1]
```diff
@@ -31,7 +31,7 @@ def get_compression_features(N, features, n_pca, partitions, landmarks):
     if n_pca == None:
         n_pca = min(N, features)
     if n_pca > 100:
-        n_pca = 100
+        n_pca = 100 # hello
```
Comment (scottgigante) [inline multiscale_phate/compress.py:34]: What's going on here?

File: multiscale_phate/embed.py [hunk 2]
```diff
@@ -1,4 +1,4 @@
-import numpy as np
+import numpy as np # loading
```
Comment (scottgigante) [inline multiscale_phate/embed.py:1]: What's this comment about

File: multiscale_phate/embed.py [hunk 3]
```diff
@@ -68,6 +68,7 @@ def compute_gradient(Xs, merges):
         Description of returned object.
 
     """
+    print("Computing gradient...")
```
Comment (scottgigante) [inline multiscale_phate/embed.py:71]: Use `tasklogger.log_info`

File: multiscale_phate/multiscale_phate.py [hunk 4]
```diff
@@ -119,42 +139,39 @@ def transform(self, X):
             Description of returned object.
 
         """
-        if utils.hash_object(X) == self.hash:
```
Comment (scottgigante) [inline multiscale_phate/multiscale_phate.py:122]: What happened to the out of sample stuff?

File: multiscale_phate/multiscale_phate.py [hunk 5]
```diff
@@ -119,42 +139,39 @@ def transform(self, X):
...
+        type
+            Description of returned object.
 
-        return hp_embedding, cluster_viz, sizes_viz, vis_tree
+        """
+        return visualize.build_condensation_tree(
+            self.data_pca, self.diff_op, self.NxTs, self.merges, self.Ps
+        )
 
-    def fit_transform(self, X):
```
Comment (scottgigante) [inline multiscale_phate/multiscale_phate.py:157]: I'd recommend keeping a fit_transform which just calls fit and then transform with default params

File: multiscale_phate/visualize.py [hunk 6]
```diff
@@ -102,4 +137,4 @@ def build_condensation_tree(data_pca, diff_op, NxT, merged_list, Ps):
 
         tree = np.vstack((embeddings))
 
-    return tree
+    return tree
```
Comment (scottgigante) [inline multiscale_phate/visualize.py:140]: Missing newline. Easiest fix for this is to run `./autoblack.sh` in the root directory to automatically reformat

File: README.md [hunk 7]
```diff
@@ -8,36 +8,37 @@ Multiscale_PHATE
...
-# Plot condensation tree
-scprep.plot.scatter3d(tree, c=tree[:,2],fontsize=16, ticks=False, label_prefix="C-PHATE", figsize=(16,12), s=20)
+Guided Tutorial
+-----------
 
-# Embed online data
-Y = np.random.normal(0.5, 1, (50, 10))
-hp_embedding, cluster_viz, sizes_viz, tree = mp_op.transform(Y)
-```
+PUT TUTORIAL LINK HERE
```
Comment (scottgigante) [inline README.md:44]: Maybe write something like "TODO."

**Review by scottgigante — CHANGES_REQUESTED** [review]

File: multiscale_phate/multiscale_phate.py [hunk 8]
```diff
@@ -60,21 +75,39 @@ def __init__(
         self.gamma = gamma
         self.knn = knn
         self.n_jobs = n_jobs
+        self.NxTs = None
```
Comment (scottgigante) [inline multiscale_phate/multiscale_phate.py:78]: Why initialize these to None?

File: multiscale_phate/multiscale_phate.py [hunk 9]
```diff
@@ -2,28 +2,43 @@
...
+        number of nearest neighbors on which to build kernel for
+        diffusion potential calculation.
+    n_jobs : integer, optional, default: 1
+        The number of jobs to use for the computation.
+        If -1 all CPUs are used. If 1 is given, no parallel computing code is
+        used at all, which is useful for debugging.
+        For n_jobs below -1, (n_cpus + 1 + n_jobs) are used. Thus for
+        n_jobs = -2, all CPUs but one are used
 
     Attributes
```
Comment (scottgigante) [inline multiscale_phate/multiscale_phate.py:43]: Attributes list should document what all the following are:
```
        self.NxTs = None
        self.Xs = None
        self.Ks = None
        self.merges = None
        self.Ps = None
        self.diff_op = None
        self.data_pca = None
        self.pca_op = None
        self.partition_clusters = None
        self.dp_pca = None
        self.epsilon = None
        self.merge_threshold = None
        self.gradient = None
        self.levels = None
```

File: multiscale_phate/multiscale_phate.py [hunk 10]
```diff
@@ -60,21 +75,39 @@ def __init__(
...
+        self.merge_threshold = None
+        self.gradient = None
+        self.levels = None
+
+
         super().__init__()
 
     def fit(self, X):
-        """Short summary.
+        """Builds a Diffusion Condensation tree to the data and computes
```
Comment (scottgigante) [inline multiscale_phate/multiscale_phate.py:97]: First line of docstring should be short and fit on one line. Detailed description can follow after two linebreaks. e.g.

```
"""Compute Multiscale PHATE on data.

Builds a condensation tree on data and computes ideal
levels of granularity via gradient analysis.

Parameters
...
"""
```

File: multiscale_phate/multiscale_phate.py [hunk 11]
```diff
@@ -60,21 +75,39 @@ def __init__(
...
+            input data with `n_samples` samples and `n_dimensions`
+            dimensions. Accepted data types: `numpy.ndarray`,
+            `scipy.sparse.spmatrix` and `pd.DataFrame`.
 
         Returns
         -------
-        type
-            Description of returned object.
-
+       multiscale_phate_operator : Multiscale_PHATE
```
Comment (scottgigante) [inline multiscale_phate/multiscale_phate.py:109]: Whitespace should match level of `Returns`

File: multiscale_phate/multiscale_phate.py [hunk 12]
```diff
@@ -60,21 +75,39 @@ def __init__(
...
+        The estimator object
```
Comment (scottgigante) [inline multiscale_phate/multiscale_phate.py:110]: Whitespace should be indented by 4

File: multiscale_phate/multiscale_phate.py [hunk 13]
```diff
@@ -103,70 +136,119 @@ def fit(self, X):
...
+            the granularity of visualization_level
         """
-        if utils.hash_object(X) == self.hash:
-            NxTs = self.NxTs
-            Xs = self.Xs
-            Ks = self.Ks
-            merges = self.merges
-            Ps = self.Ps
-            data_pca = self.data_pca
+        if visualization_level is None and cluster_level is None and coarse_cluster_level is None and coarse_cluster is None:
```
Comment (scottgigante) [inline multiscale_phate/multiscale_phate.py:175]: Better to do 

```
if visualization_level is None:
    visualization_level = self.levels[2]
if cluster_level is None:
    cluster_level = self.levels[-2]
if coarse_cluster_level is None and coarse_cluster is None:
    ...
else:
    ...
```

File: multiscale_phate/multiscale_phate.py [hunk 14]
```diff
@@ -103,70 +136,119 @@ def fit(self, X):
...
-        hp_embedding, cluster_viz, sizes_viz = visualize.build_visualization(
-            Xs, NxTs, merges
-        )
+    def build_tree(self):
+        """Computes and returns a tree from the Diffusion Condensation process.
 
-        vis_tree = visualize.build_condensation_tree(
-            data_pca, self.diff_op, NxTs, merges, Ps
-        )
+        Parameters
```
Comment (scottgigante) [inline multiscale_phate/multiscale_phate.py:192]: Remove Parameters section if empty

File: multiscale_phate/multiscale_phate.py [hunk 15]
```diff
@@ -103,70 +136,119 @@ def fit(self, X):
...
+        ----------
+        None
 
-        return hp_embedding, cluster_viz, sizes_viz, vis_tree
+        Returns
+        -------
+        embedding : array, shape=[number of aggregated points
```
Comment (scottgigante) [inline multiscale_phate/multiscale_phate.py:198]: Keep topline description on one line. `shape=[n_points_aggregated, 3]` will do

File: multiscale_phate/multiscale_phate.py [hunk 16]
```diff
@@ -103,70 +136,119 @@ def fit(self, X):
...
+        ideal levels of granularity via gradient analysis before
+        computing a Multiscale PHATE embedding at a ideal visualization
+        and clustering granularities.
+
+        Parameters
+        ----------
+        X : array, shape=[n_samples, n_features]
+            input data with `n_samples` samples and `n_dimensions`
+            dimensions. Accepted data types: `numpy.ndarray`,
+            `scipy.sparse.spmatrix` and `pd.DataFrame`.
```
Comment (scottgigante) [inline multiscale_phate/multiscale_phate.py:218]: Newline between Parameters and Returns

File: multiscale_phate/multiscale_phate.py [hunk 17]
```diff
@@ -103,70 +136,119 @@ def fit(self, X):
...
-        X : type
-            Description of parameter `X`.
+        cluster_level : int
+            Resolution of Diffusion Condensation tree to visualize clusters.
 
         Returns
         -------
-        type
-            Description of returned object.
+        clusters_tree : array shape = array, shape=[number of aggregated points
```
Comment (scottgigante) [inline multiscale_phate/multiscale_phate.py:244]: Keep this to one line

File: multiscale_phate/multiscale_phate.py [hunk 18]
```diff
@@ -103,70 +136,119 @@ def fit(self, X):
...
+        across granularities, 1]
```
Comment (scottgigante) [inline multiscale_phate/multiscale_phate.py:245]: Is it array-like [n, 1] or is it list-like [n]?

**Review by scottgigante — CHANGES_REQUESTED** [review]

File: .idea/Multiscale_PHATE.iml [hunk 19]
```diff
@@ -0,0 +1,15 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
[the remark this answers is not in the document]
Comment (scottgigante) [reply .idea/Multiscale_PHATE.iml:1]: Bump

File: multiscale_phate/multiscale_phate.py (hunk 8 above)
Comment (scottgigante) [reply multiscale_phate/multiscale_phate.py:129]: Bump

File: multiscale_phate/multiscale_phate.py (hunk 9 above)
Comment (scottgigante) [reply multiscale_phate/multiscale_phate.py:78]: You should describe these -- they are the primary attributes of your estimator. I have no idea what NxTs means, for example

File: multiscale_phate/multiscale_phate.py (hunk 12 above)
Comment (scottgigante) [reply multiscale_phate/multiscale_phate.py:159]: This is still not indented correctly -- `The estimator object` should be indented by 4 more spaces.

File: multiscale_phate/multiscale_phate.py [hunk 20]
```diff
@@ -102,71 +184,110 @@ def fit(self, X):
...
 
-    def transform(self, X):
-        """Short summary.
+        self.gradient = embed.compute_gradient(self.Xs, self.merges)
+        self.levels = embed.get_levels(self.gradient)
 
+        return self.levels
+
+    def transform(self, visualization_level=None, cluster_level=None, coarse_cluster_level = None, coarse_cluster = None, repulse = False):
+        """Short summary.
```
Comment (scottgigante) [inline multiscale_phate/multiscale_phate.py:196]: What does `transform` do

**Review by scottgigante — CHANGES_REQUESTED** [review]

**Review by scottgigante — APPROVED** [review]
